### DISCUSSION TITLE
Race condition when using .live.debounce modifiers

### DISCUSSION THREAD

**Discussion by efremov-ie** [livewire/livewire#8466] [7. Bugs (archived)]
### Livewire version

v3.4.11

### Laravel version

v10.48

### Which PHP version are you using?

PHP 8.2

### Steps To Reproduce

1. Create a table with 10-20 checkboxes 
2. Each checkbox has an attribute like wire:model.live.debounce.2000="values.{elementId}".
3. activate the mode of slow 3g internet connection for the simulation
4. Select multiple checkboxes with the user's normal speed
5. After sending a request to the server, only the first 1-2 checkboxes are selected, the state of the rest is reset
![CPT2405211342-244x263](https://github.com/livewire/livewire/assets/13541867/ded8314b-5030-440e-9b53-e4d747ee6c3f)

### Repository or Wirebox URL

-

### Additional information

Expected: livewire debounce will only send the latest request, but will merge all previous requests together. This way the user will be able to select many options, but only 1 request will be sent to synchronise all selected items at once. This is also how debounce works now in Lodash JS

Code example:
Component: 
```
<?php
namespace App\Livewire\Moment;

use Livewire\Component;

class ValuesTable extends Component
{
    public array $values = [];

    .....

    public function render()
    {
        return view('path.to.values-table');
    }
}
```

Template:
```
      <table>
          @foreach($values as $value)
              <tr wire:key="{{ Str::random() }}">
                  <td>
                      <input type="checkbox" wire:model.live.debounce.2000="values.{{ $value['id'] }}" />
                  </td>
                  <td>{{ $value['title'] }}</td>
              </tr>
          @endforeach
      </table>
```

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by WeiChun0911** [comment]
I believe this is not a bug; it works as expected.

This is because you have set debounce for each checkbox. 
The debounce feature will work if you click the same checkbox multiple times.

However, if you click different checkboxes, it will fire multiple update events, which can cause a race condition, since there will be multiple AJAX requests trying to modify `public array $values = [];.`

I also encountered this issue, but I still don't know how to resolve it properly.

**Comment by lrljoe** [comment]
As said elsewhere, it's not a bug per-se

Livewire needs to send the update at some point, it doesn't know whether the user is going to check another box or not, so it can't wait forever.

You could keep the updates deferred, and use AlpineJS to look for either
Mouse leaving the parent element 
No mouse movement/clicks for x ms
Then send the updates

It'd also be pretty clean to use AlpineJs and entangle for this, which would give you more flexibility

**Reply by efremov-ie** [reply]
This is a great idea; I’ll try it soon and write here about how well it works.

In my opinion, it’s strange that Livewire doesn’t account for HTML5 functionality in this case. If I have a list of checkboxes, they should work together, handling race conditions properly. However, if it’s a radio button or a single checkbox, the behavior should work as it currently does.

**Comment by MariaMozgunova** [comment]
I also found that if no server-side update is needed at that moment — only the front end needs to reflect the change — then working with and modifying the property directly on the `$wire` object works well. The related updates on the front end are handled correctly by Livewire 3, and since it's just a client-side mutation (not a dispatch/commit), it doesn't fire its own request that can race against other fields' updates.

In my case, I had a bunch of unrelated fields (dates, radios, checkboxes, selects) that each needed to append a snippet of text to a shared notes field whenever they changed:

```blade
<input type="text" wire:model="field1"
    onblur="addTextToNotes('Updated: ' + this.value + '\r\n')">

<input type="radio" wire:model="field2" value="1"
    onchange="addTextToNotes('Updated: ' + this.value + '\r\n')">
```

Originally `addTextToNotes` dispatched an event, and the component handled it server-side:

```js
function addTextToNotes(note) {
    Livewire.dispatch('addTextToNotes', {value: note});
}
```

That's its own commit — if two of these fields changed in quick succession, two overlapping requests would go out, and whichever response arrived last would win, sometimes silently reverting whatever the other field had just changed.

Switching it to mutate `$wire` directly fixed it, since there's no longer a second request to race against the field's own (deferred) commit:

```js
function addTextToNotes(note) {
    let wire = Livewire.getByName('my-component-name')[0];
    if (!wire) return;

    wire.notes += note;
}
```

`Livewire.getByName('find-customer')[0]` gets the `$wire` instance for that component from plain global JS (equivalent to `$wire` inside an Alpine expression). The note text just accumulates in Livewire's reactive state and gets sent along whenever the next real commit happens — no separate round trip, so nothing to race.
